transport/grpc/outbound_test.go

```diff
@@ -200,3 +202,84 @@ func TestCallStreamWithInvalidPeer(t *testing.T) {
 		}.Error(),
 	)
 }
+
+func TestCallSvcMatch(t *testing.T) {
+	tests := []struct {
+		msg         string
+		serverPort  string
+		headerKey   string
+		headerValue string
+		wantErr     bool
+	}{
+		{
+			msg:         "call service match success",
+			serverPort:  "15213",
+			headerKey:   ServiceHeader,
+			headerValue: "Service",
+		},
+		{
+			msg:         "call service match failed",
+			serverPort:  "15619",
+			headerKey:   ServiceHeader,
+			headerValue: "ThisIsWrongSvcName",
+			wantErr:     true,
+		},
+		{
+			msg:        "no service name response header",
+			serverPort: "15640",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.msg, func(t *testing.T) {
+			server := grpc.NewServer(
+				grpc.CustomCodec(customCodec{}),
+				grpc.UnknownServiceHandler(func(srv interface{}, stream grpc.ServerStream) error {
+					responseWriter := newResponseWriter()
+					defer responseWriter.Close()
+
+					if tt.headerKey != "" {
+						responseWriter.AddSystemHeader(tt.headerKey, tt.headerValue)
+					}
+
+					// Send the response attributes back and end the stream.
+					if sendErr := stream.SendMsg(responseWriter.Bytes()); sendErr != nil {
+						// We couldn't send the response.
+						return sendErr
+					}
+					if responseWriter.md != nil {
+						stream.SetTrailer(responseWriter.md)
+					}
+					return nil
+				}),
+			)
+			listener, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%s", tt.serverPort))
+			require.NoError(t, err)
+			go func() {
+				err := server.Serve(listener)
+				require.NoError(t, err)
+			}()
+			defer server.Stop()
+
+			grpcTransport := NewTransport()
+			out := grpcTransport.NewSingleOutbound(listener.Addr().String())
+			require.NoError(t, grpcTransport.Start())
+			require.NoError(t, out.Start())
+			defer grpcTransport.Stop()
+			defer out.Stop()
+
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+			defer cancel()
+			req := &transport.Request{
+				Service:   "Service",
+				Procedure: "Hello",
+				Body:      bytes.NewReader([]byte("world")),
+			}
+			_, err = out.Call(ctx, req)
+			if tt.wantErr {
+				assert.Contains(t, err.Error(), "does not match")
```

Comment:
Do a `require.Error(t, err)` before calling `err.Error()` otherwise if an
error was expected but was nil, this code will panic.

Reply:
added.